File: 124.二叉树中的最大路径和.go
import "math"

/*
 * @lc app=leetcode.cn id=124 lang=golang
 *
 * [124] 二叉树中的最大路径和
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func maxPathSum(root *TreeNode) int {
	if root == nil {
		return 0
	}

	max := math.MinInt32
	_ = oneSideMax(root, &max)
	return max
}

func oneSideMax(root *TreeNode, maxSum *int) int {
	if root == nil {
		return 0
	}

	left := oneSideMax(root.Left, maxSum)
	right := oneSideMax(root.Right, maxSum)

	curMax := max(max(left+root.Val, right+root.Val), root.Val)
	*maxSum = max(*maxSum, max(curMax, left+right+root.Val))
	return curMax
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

// @lc code=end

File: 124.二叉树中的最大路径和.java
import javax.swing.tree.TreeNode;

/*
 * @lc app=leetcode.cn id=124 lang=java
 *
 * [124] 二叉树中的最大路径和
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private int maxSum = Integer.MIN_VALUE;

    public int maxPathSum(TreeNode root) {
        if (root == null) {
            return 0;
        }

        oneSideMax(root);
        return maxSum;
    }

    private int oneSideMax(TreeNode root) {
        if (root == null) {
            return 0;
        }

        int left = oneSideMax(root.left);
        int right = oneSideMax(root.right);

        int curMax = Math.max(Math.max(left + root.val, right + root.val), root.val);
        maxSum = Math.max(maxSum, Math.max(curMax, left + right + root.val));
        return curMax;
    }
}
// @lc code=end

